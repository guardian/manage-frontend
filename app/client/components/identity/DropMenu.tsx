import { CSSObject } from '@emotion/core';
import { FC, useState } from 'react';
import palette from '../../colours';
import { serif } from '../../styles/fonts';
interface DropMenuProps {
	title: string;
	color: string;
}

const styles = {
	root: {
		borderColor: palette.neutral['4'],
		borderTop: `1px solid ${palette.neutral['6']}`,
		cursor: 'pointer',
		padding: '3px 0 12px',
		position: 'relative',
	} as CSSObject,
	header: {
		fontSize: '17px',
		fontFamily: serif,
		fontWeight: 'bold',
		lineHeight: '24px',
		textTransform: 'capitalize',
		':after': {
			content: "''",
			border: '2px solid currentColor',
			borderLeft: 'transparent',
			borderTop: 'transparent',
			boxSizing: 'content-box',
			display: 'inline-block',
			height: '5px',
			color: 'inherit',
			transform: 'translateY(-2px) rotate(45deg)',
			transition: 'transform 250ms ease-out',
			verticalAlign: 'middle',
			width: '5px',
			marginLeft: '4px',
		},
		'&.open:after': {
			transform: 'rotate(225deg)',
		},
	} as CSSObject,
};

export const DropMenu: FC<DropMenuProps> = (props) => {
	const { children, color, title } = props;
	const [open, setOpen] = useState(false);
	return (
		<div css={styles.root}>
			<div
				className={open ? 'open' : undefined}
				css={{ ...styles.header, color }}
				onClick={() => setOpen(!open)}
			>
				{title}
			</div>
			{open && children}
		</div>
	);
};
