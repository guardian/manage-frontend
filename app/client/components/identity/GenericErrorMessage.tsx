import { forwardRef } from 'react';
import palette from '../../colours';
import { errorMessageCss } from './sharedStyles';

export type GenericErrorMessageRef = HTMLDivElement;
export const GenericErrorMessage = forwardRef<GenericErrorMessageRef, {}>(
	(_, ref) => {
		return (
			<div ref={ref} css={errorMessageCss}>
				<p css={{ marginBottom: '10px' }}>
					Sorry, something went wrong!
				</p>
				<p css={{ marginBottom: '0' }}>
					<a
						css={{
							color: palette.blue.dark,
							cursor: 'pointer',
						}}
						onClick={() => window.location.reload()}
					>
						Refresh this page
					</a>
				</p>
			</div>
		);
	},
);
