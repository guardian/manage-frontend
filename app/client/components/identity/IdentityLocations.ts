import { conf } from '../../../server/config';

const url = (subdomain: string, domain: string, path?: string) =>
	`https://${subdomain}.${domain}${path ? path : ''}`;

const DOMAIN =
	typeof window !== 'undefined' && window.guardian
		? window.guardian.domain
		: conf.DOMAIN;

const IMAGE_DOMAIN =
	DOMAIN === 'theguardian.com' ? 'guim.co.uk' : 'guimcode.co.uk';

const IDAPI_URL =
	DOMAIN === 'thegulocal.com' ? '/idapicodeproxy' : url('idapi', DOMAIN);

const AVATAR_URL =
	DOMAIN === 'thegulocal.com' ? '/avatarcodeproxy' : url('avatar', DOMAIN);

const getIdentityLocations = (domain: string) => ({
	COMMUNITY_FAQS: url('www', domain, '/community-faqs'),
	CONTACT_AND_DELIVERY_HELP: url(
		'manage',
		'theguardian.com',
		'/help-centre/article/i-need-to-change-my-delivery-address',
	),
	CHANGE_EMAIL: url('profile', domain, '/account/edit'),
	RESET_PASSWORD: url('profile', domain, '/reset'),
	MANAGE_JOB_ALERTS: url('jobs', domain, '/your-jobs/?ActiveSection=JbeList'),
	VERIFY_EMAIL: url('profile', domain, '/verify-email'),
	IDAPI: IDAPI_URL,
	AVATAR: AVATAR_URL,
	AVATAR_USER_IMAGES: url('avatar', IMAGE_DOMAIN, '/user'),
	DELETE_ACCOUNT: url('profile', domain, '/delete'),
});

export const IdentityLocations = getIdentityLocations(DOMAIN);
