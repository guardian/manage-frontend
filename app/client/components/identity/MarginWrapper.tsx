import { FC } from "react";
import { minWidth } from "../../styles/breakpoints";

export const MarginWrapper: FC<{}> = ({ children }) => {
  return (
    <div
      css={{
        boxSizing: "border-box",
        margin: "0 auto",
        maxWidth: "620px",
        [minWidth.desktop]: {
          maxWidth: "100%",
        },
      }}
    >
      {children}
    </div>
  );
};
