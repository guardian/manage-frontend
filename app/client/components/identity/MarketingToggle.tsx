import React, { FC } from "react";
import { sans } from "../../styles/fonts";
import { css } from "@emotion/core";
import { ToggleSwitch } from "./ToggleSwitch";

interface MarketingToggleProps {
  id: string;
  description?: string;
  title?: string;
  selected?: boolean;
  onClick: (id: string) => {};
}

const standardText = {
  fontFamily: sans,
  fontSize: "14px",
  lineHeight: "1.333"
};

const getDescription = (description: MarketingToggleProps["description"]) => (
  <p
    css={[
      standardText,
      {
        margin: "0",
        padding: "2.88px 90px 0 0"
      }
    ]}
  >
    {description}
  </p>
);

export const MarketingToggle: FC<MarketingToggleProps> = (props) => {
  const { id, description, selected, title, onClick } = props;
  return (
    <div
      onClick={(e) => {
        // Checkboxes inside labels will trigger click events twice.
        // Ignore the input click event
        if (e.target instanceof Element && e.target.nodeName === "INPUT") {
          return;
        }
        onClick(id);
      }}
      css={[
        standardText,
        {
          marginTop: "12px",
          position: "relative"
        }
      ]}
    >
      <div css={{ left: 0 }}>
        <ToggleSwitch
          cssOverrides={css`
            label {
              ${standardText};
              padding-right: 5px;
              font-weight: bold;
            }
          `}
          label={title}
          labelPosition="left"
          labelId={id}
          checked={!!selected}
          onClick={(_) => {
            return;
          }}
        />
      </div>
      {description && getDescription(description)}
    </div>
  );
};
