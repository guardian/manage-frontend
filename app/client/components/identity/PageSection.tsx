import { FC, ReactNode } from 'react';
import palette from '../../colours';
import { minWidth } from '../../styles/breakpoints';
import { sans } from '../../styles/fonts';

interface PageSectionProps {
	title?: string;
	description?: string | ReactNode;
	subtext?: string;
}

const getTitle = (title: PageSectionProps['title']) => (
	<h2
		css={{
			fontSize: '17px',
			lineHeight: '24px',
			fontWeight: 'bold',
			margin: '0 0 6px 0',
		}}
	>
		{title}
	</h2>
);

const getDescription = (description: PageSectionProps['description']) => (
	<p
		css={{
			fontFamily: sans,
			fontSize: '14px',
			marginBottom: '8px',
		}}
	>
		{description}
	</p>
);

const getSubtext = (subtext: PageSectionProps['subtext']) => (
	<p
		css={{
			fontSize: '13px',
			lineHeight: '18px',
			fontFamily: sans,
			color: palette.neutral['3'],
			marginBottom: '12px',
			marginTop: '0',
		}}
	>
		{subtext}
	</p>
);

export const PageSection: FC<PageSectionProps> = (props) => {
	const { children, description, title, subtext } = props;
	return (
		<div
			css={{
				[minWidth.desktop]: {
					display: 'flex',
				},
			}}
		>
			<div
				css={{
					[minWidth.desktop]: {
						paddingRight: '100px',
						boxSizing: 'content-box',
						flexBasis: '220px',
						minWidth: '220px',
					},
				}}
			>
				{title && getTitle(title)}
				{description && getDescription(description)}
				{subtext && getSubtext(subtext)}
			</div>
			<div
				css={{
					[minWidth.desktop]: {
						maxWidth: '460px',
						flexGrow: 1,
					},
				}}
			>
				{children}
			</div>
		</div>
	);
};
