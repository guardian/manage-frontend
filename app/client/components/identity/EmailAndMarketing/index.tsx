import React, { useEffect, useState } from "react";
import { ProductDetail } from "../../../../shared/productResponse";
import { GROUPED_PRODUCT_TYPES } from "../../../../shared/productTypes";
import { allProductsDetailFetcher } from "../../../productUtils";
import { NAV_LINKS } from "../../nav/navConfig";
import { PageContainer } from "../../page";
import { Spinner } from "../../spinner";
import { WithStandardTopMargin } from "../../WithStandardTopMargin";
import {
  GenericErrorMessage,
  GenericErrorMessageRef
} from "../GenericErrorMessage";
import { ConsentOptions, Users } from "../identity";
import { IdentityLocations } from "../IdentityLocations";
import { Lines } from "../Lines";
import { ConsentOption } from "../models";
import { Actions, useConsentOptions } from "../useConsentOptions";
import { ConsentSection } from "./ConsentSection";
import { EmailSettingsSection } from "./EmailSettingsSection";
import { NewsletterSection } from "./NewsletterSection";
import { OptOutSection } from "./OptOutSection";

const EmailAndMarketing = (_: { path?: string }) => {
  const { options, error, subscribe, unsubscribe, unsubscribeAll } = Actions;
  const [email, setEmail] = useState<string>("");
  const [removed, setRemoved] = useState(false);
  const [state, dispatch] = useConsentOptions();

  const toggleSubscription = async (id: string) => {
    const option = ConsentOptions.findById(state.options, id);
    try {
      if (option === undefined) {
        throw Error("Id not found");
      }
      if (option.subscribed) {
        await ConsentOptions.unsubscribe(option);
        dispatch(unsubscribe(id));
      } else {
        await ConsentOptions.subscribe(option);
        dispatch(subscribe(id));
      }
    } catch (e) {
      dispatch(error(e));
    }
  };

  const setRemoveAllEmailConsents = async () => {
    try {
      await ConsentOptions.unsubscribeAll();
      setRemoved(true);
      dispatch(unsubscribeAll());
    } catch (e) {
      dispatch(error(e));
    }
  };

  useEffect(() => {
    const makeInitialAPICalls = async () => {
      try {
        const user = await Users.getCurrentUser();
        if (!user.validated) {
          window.location.assign(IdentityLocations.VERIFY_EMAIL);
          return;
        }
        const productDetailsResponse = await allProductsDetailFetcher();
        const productDetails: ProductDetail[] = await productDetailsResponse.json();
        const consentOptions = await ConsentOptions.getAll();
        const consentsWithFilteredSoftOptIns = consentOptions.filter(
          (consent: ConsentOption) =>
            consent.isProduct
              ? productDetails.some(productDetail => {
                  const groupedProductType =
                    GROUPED_PRODUCT_TYPES[productDetail.mmaCategory];
                  const specificProductType = groupedProductType.mapGroupedToSpecific(
                    productDetail
                  );
                  return specificProductType.softOptInIDs.includes(consent.id);
                })
              : true
        );

        setEmail(user.primaryEmailAddress);
        dispatch(options(consentsWithFilteredSoftOptIns));
      } catch (e) {
        dispatch(error(e));
      }
    };
    makeInitialAPICalls();
  }, []);

  const newsletters = ConsentOptions.newsletters(state.options);
  const consents = ConsentOptions.consents(state.options);
  const loading = newsletters.length === 0 && consents.length === 0;

  const errorRef = React.createRef<GenericErrorMessageRef>();

  useEffect(() => {
    if (state.error && errorRef.current) {
      window.scrollTo(0, errorRef.current.offsetTop - 20);
    }
  }, [state.error]);

  const errorMessage = (
    <WithStandardTopMargin>
      <GenericErrorMessage ref={errorRef} />
    </WithStandardTopMargin>
  );

  const content = (
    <>
      <WithStandardTopMargin>
        <NewsletterSection
          newsletters={newsletters}
          clickHandler={toggleSubscription}
        />
      </WithStandardTopMargin>
      <WithStandardTopMargin>
        <Lines n={4} />
      </WithStandardTopMargin>
      <WithStandardTopMargin>
        <ConsentSection consents={consents} clickHandler={toggleSubscription} />
      </WithStandardTopMargin>
      <WithStandardTopMargin>
        <Lines n={1} />
      </WithStandardTopMargin>
      <WithStandardTopMargin>
        <EmailSettingsSection
          email={email}
          actionHandler={setRemoveAllEmailConsents}
          removed={removed}
        />
      </WithStandardTopMargin>
      <WithStandardTopMargin>
        <Lines n={4} />
      </WithStandardTopMargin>
      <WithStandardTopMargin>
        <OptOutSection consents={consents} clickHandler={toggleSubscription} />
      </WithStandardTopMargin>
    </>
  );

  const loader = (
    <WithStandardTopMargin>
      <Spinner loadingMessage="Loading your email preferences..." />
    </WithStandardTopMargin>
  );

  return (
    <PageContainer
      selectedNavItem={NAV_LINKS.emailPrefs}
      pageTitle="Emails &amp; marketing"
    >
      {state.error ? errorMessage : null}
      {loading ? (!state.error ? loader : null) : content}
    </PageContainer>
  );
};

export default EmailAndMarketing;
