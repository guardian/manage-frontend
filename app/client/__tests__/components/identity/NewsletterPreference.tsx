import { cleanup, fireEvent, render } from '@testing-library/react';
import React from 'react';
import { create } from 'react-test-renderer';
import { NewsletterPreference } from '../../../components/identity/NewsletterPreference';

afterEach(cleanup);

describe('NewsletterPreference component', () => {
	const input = {
		id: '13',
		description: 'Test description',
		frequency: 'Test frequency',
		title: 'Test title',
		clickHandler: jest.fn(),
		identityName: 'Test-13',
	};
	it('renders correctly and displays marketing information', () => {
		const rendered = create(
			<NewsletterPreference
				id={input.id}
				description={input.description}
				frequency={input.frequency}
				identityName={input.identityName}
				title={input.title}
				onClick={input.clickHandler}
			/>,
		);
		expect(rendered.toJSON()).toMatchSnapshot();
	});

	it('will select the checkbox when the selected prop is passed', () => {
		const rendered = create(
			<NewsletterPreference
				id={input.id}
				description={input.description}
				frequency={input.frequency}
				identityName={input.identityName}
				title={input.title}
				selected={true}
				onClick={input.clickHandler}
			/>,
		);
		expect(rendered.toJSON()).toMatchSnapshot();
	});

	it('will call the click handler when it is clicked', () => {
		const { getByText } = render(
			<NewsletterPreference
				id={input.id}
				description={input.description}
				frequency={input.frequency}
				identityName={input.identityName}
				title={input.title}
				onClick={input.clickHandler}
			/>,
		);
		fireEvent.click(getByText(input.title));
		expect(input.clickHandler).toHaveBeenCalledTimes(1);
		expect(input.clickHandler).toHaveBeenLastCalledWith(input.id);
	});
});
