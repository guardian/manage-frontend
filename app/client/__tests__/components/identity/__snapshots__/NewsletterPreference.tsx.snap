// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsletterPreference component renders correctly and displays marketing information 1`] = `
<div
  css={
    Array [
      Object {
        "fontFamily": "\\"Guardian Text Sans Web\\", \\"Helvetica Neue\\", Helvetica, Arial, \\"Lucida Grande\\", sans-serif",
        "fontSize": "14px",
      },
      Object {
        "lineHeight": "1.333",
        "marginTop": "12px",
        "paddingLeft": "30px",
        "position": "relative",
      },
    ]
  }
  onClick={[Function]}
>
  <div
    css={
      Object {
        "left": 0,
        "position": "absolute",
      }
    }
  >
    <label
      css={
        Object {
          ":hover .checkbox": Object {
            "boxShadow": "0 0 0 3px #ececec",
          },
          "cursor": "pointer",
          "display": "flex",
          "maxWidth": undefined,
          "userSelect": "none",
        }
      }
    >
      <div
        aria-checked={false}
        className="checkbox"
        css={
          Object {
            ":focus": Object {
              "boxShadow": "0 0 0 3px #ffe500",
            },
            "background": "#ffffff",
            "border": "1px solid #999999",
            "display": "inline-block",
            "height": "18px",
            "margin": "3px",
            "marginRight": "10px",
            "minWidth": "18px",
            "outline": 0,
            "position": "relative",
            "transition": "all .2s ease-in-out",
          }
        }
        onKeyPress={[Function]}
        role="checkbox"
        tabIndex={0}
      >
        <input
          checked={false}
          css={
            Object {
              "opacity": 0,
              "overflow": "hidden",
              "position": "absolute",
              "zIndex": -999999,
            }
          }
          onChange={[Function]}
          tabIndex={-1}
          type="checkbox"
        />
        <div
          css={
            Object {
              "height": "6px",
              "left": "3px",
              "position": "absolute",
              "textAlign": "right",
              "top": "5px",
              "transform": "rotate(-45deg)",
              "width": "12px",
            }
          }
        >
          <div
            css={
              Object {
                "borderColor": "#ffffff",
                "borderStyle": "solid",
                "borderWidth": "0 0 2px 2px",
                "height": "6px",
                "transition": "all .2s ease-in-out",
                "transitionDelay": ".1s",
                "width": "2px",
              }
            }
          />
        </div>
      </div>
      <span />
    </label>
  </div>
  <p
    css={
      Array [
        Object {
          "fontFamily": "\\"Guardian Text Sans Web\\", \\"Helvetica Neue\\", Helvetica, Arial, \\"Lucida Grande\\", sans-serif",
          "fontSize": "14px",
        },
        Object {
          "cursor": "pointer",
          "fontFamily": "\\"Guardian Text Sans Web\\", \\"Helvetica Neue\\", Helvetica, Arial, \\"Lucida Grande\\", sans-serif",
          "fontSize": "14px",
          "fontWeight": "bold",
          "lineHeight": "22px",
          "margin": "0",
        },
      ]
    }
  >
    Test title
  </p>
  <p
    css={
      Object {
        "padding": "2.88px 0 0 0",
      }
    }
  >
    Test description
  </p>
  <p
    css={
      Object {
        "fontSize": "12px",
        "lineHeight": "16px",
        "margin": "3px 0 0 0",
        "opacity": 0.75,
      }
    }
  >
    <span
      css={
        Object {
          "display": "inline-block",
          "marginRight": "8px",
          "verticalAlign": "middle",
        }
      }
    >
      <svg
        css={
          Object {
            "fill": "#dcdcdc",
          }
        }
        height="11px"
        viewBox="0 0 11 11"
        width="11px"
      >
        <path
          d="M5.4 0C2.4 0 0 2.4 0 5.4s2.4 5.4 5.4 5.4 5.4-2.4 5.4-5.4S8.4 0 5.4 0zm3 6.8H4.7V1.7h.7L6 5.4l2.4.6v.8z"
        />
      </svg>
    </span>
    Test frequency
  </p>
</div>
`;

exports[`NewsletterPreference component will select the checkbox when the selected prop is passed 1`] = `
<div
  css={
    Array [
      Object {
        "fontFamily": "\\"Guardian Text Sans Web\\", \\"Helvetica Neue\\", Helvetica, Arial, \\"Lucida Grande\\", sans-serif",
        "fontSize": "14px",
      },
      Object {
        "lineHeight": "1.333",
        "marginTop": "12px",
        "paddingLeft": "30px",
        "position": "relative",
      },
    ]
  }
  onClick={[Function]}
>
  <div
    css={
      Object {
        "left": 0,
        "position": "absolute",
      }
    }
  >
    <label
      css={
        Object {
          ":hover .checkbox": Object {
            "boxShadow": "0 0 0 3px #ececec",
          },
          "cursor": "pointer",
          "display": "flex",
          "maxWidth": undefined,
          "userSelect": "none",
        }
      }
    >
      <div
        aria-checked={true}
        className="checkbox"
        css={
          Object {
            ":focus": Object {
              "boxShadow": "0 0 0 3px #ffe500",
            },
            "background": "#22834D",
            "border": undefined,
            "display": "inline-block",
            "height": "18px",
            "margin": "3px",
            "marginRight": "10px",
            "minWidth": "18px",
            "outline": 0,
            "position": "relative",
            "transition": "all .2s ease-in-out",
          }
        }
        onKeyPress={[Function]}
        role="checkbox"
        tabIndex={0}
      >
        <input
          checked={true}
          css={
            Object {
              "opacity": 0,
              "overflow": "hidden",
              "position": "absolute",
              "zIndex": -999999,
            }
          }
          onChange={[Function]}
          tabIndex={-1}
          type="checkbox"
        />
        <div
          css={
            Object {
              "height": "6px",
              "left": "3px",
              "position": "absolute",
              "textAlign": "right",
              "top": "5px",
              "transform": "rotate(-45deg)",
              "width": "12px",
            }
          }
        >
          <div
            css={
              Object {
                "borderColor": "#ffffff",
                "borderStyle": "solid",
                "borderWidth": "0 0 2px 2px",
                "height": "6px",
                "transition": "all .2s ease-in-out",
                "transitionDelay": ".1s",
                "width": "12px",
              }
            }
          />
        </div>
      </div>
      <span />
    </label>
  </div>
  <p
    css={
      Array [
        Object {
          "fontFamily": "\\"Guardian Text Sans Web\\", \\"Helvetica Neue\\", Helvetica, Arial, \\"Lucida Grande\\", sans-serif",
          "fontSize": "14px",
        },
        Object {
          "cursor": "pointer",
          "fontFamily": "\\"Guardian Text Sans Web\\", \\"Helvetica Neue\\", Helvetica, Arial, \\"Lucida Grande\\", sans-serif",
          "fontSize": "14px",
          "fontWeight": "bold",
          "lineHeight": "22px",
          "margin": "0",
        },
      ]
    }
  >
    Test title
  </p>
  <p
    css={
      Object {
        "padding": "2.88px 0 0 0",
      }
    }
  >
    Test description
  </p>
  <p
    css={
      Object {
        "fontSize": "12px",
        "lineHeight": "16px",
        "margin": "3px 0 0 0",
        "opacity": 0.75,
      }
    }
  >
    <span
      css={
        Object {
          "display": "inline-block",
          "marginRight": "8px",
          "verticalAlign": "middle",
        }
      }
    >
      <svg
        css={
          Object {
            "fill": "#dcdcdc",
          }
        }
        height="11px"
        viewBox="0 0 11 11"
        width="11px"
      >
        <path
          d="M5.4 0C2.4 0 0 2.4 0 5.4s2.4 5.4 5.4 5.4 5.4-2.4 5.4-5.4S8.4 0 5.4 0zm3 6.8H4.7V1.7h.7L6 5.4l2.4.6v.8z"
        />
      </svg>
    </span>
    Test frequency
  </p>
</div>
`;
